src/Tmds.Ssh/KerberosCredential.cs

```diff
@@ -0,0 +1,339 @@
+        // client supplied username as the target user. The Kerberos principal credential is only used for the
+        // authentication stage that happens next.
+        string userName = settings.UserName;
+        logger.AuthenticationMethodGssapiWithMic(userName, _kerberosCredential.UserName, spn, _delegateCredential);
+
+        bool isOidSuccess = await TryStageOid(connection, logger, userName, ct).ConfigureAwait(false);
+        if (!isOidSuccess)
+        {
+            return false;
+        }
+
+        var negotiateOptions = new NegotiateAuthenticationClientOptions()
+        {
+            AllowedImpersonationLevel = _delegateCredential ? TokenImpersonationLevel.Delegation : TokenImpersonationLevel.Impersonation,
+            Credential = _kerberosCredential,
+            Package = "Kerberos",
+            // While only Sign is needed we need to set EncryptAndSign for
+            // Windows client support. Sign only will pass in SECQOP_WRAP_NO_ENCRYPT
+            // to MakeSignature which fails.
+            // https://github.com/dotnet/runtime/issues/103461
+            RequiredProtectionLevel = ProtectionLevel.EncryptAndSign,
+            // While RFC states this should be set to "false", Win32-OpenSSH
+            // fails if it's not true. I'm unsure if openssh-portable on Linux
+            // will fail in the same way or not.
+            RequireMutualAuthentication = true,
+            TargetName = spn,
+        };
+
+        using var authContext = new NegotiateAuthentication(negotiateOptions);
+        bool isAuthSuccess = await TryStageAuthentication(connection, logger, authContext, ct).ConfigureAwait(false);
+        if (!isAuthSuccess)
+        {
+            return false;
+        }
+
+        try
+        {
+            using var message = authContext.IsSigned
+                ? CreateGssapiMicData(connection.SequencePool, connectionInfo.SessionId!,  userName, authContext)
+                : CreateGssapiCompleteMessage(connection.SequencePool);
+            await connection.SendPacketAsync(message.Move(), ct).ConfigureAwait(false);
+        }
+        catch (MissingMethodException)
+        {
+            // Remove once .NET 8 is no longer the minimum and we get rid of reflection.
+            return false;
+        }
+
+        return true;
+    }
+
+    private static async Task<bool> TryStageOid(SshConnection connection, ILogger logger, string userName, CancellationToken ct)
+    {
+        {
+            using var userAuthMsg = CreateOidRequestMessage(connection.SequencePool,
+                                        userName, KRB5_OID);
+            await connection.SendPacketAsync(userAuthMsg.Move(), ct).ConfigureAwait(false);
+        }
+
+        while (true)
+        {
+            using Packet response = await connection.ReceivePacketAsync(ct).ConfigureAwait(false);
+            MessageId? mid = response.MessageId;
+
+            if (mid == MessageId.SSH_MSG_USERAUTH_BANNER)
```

Comment:
+1

There are probably other cases where this message needs to be filtered out. I created https://github.com/tmds/Tmds.Ssh/issues/192 to look into it further.

Reply:
Sounds good, I came across it when manually testing against a banner. It seems like only the first message in the exchange contains the banner. Having a generic way to handle this would be great but for now I'm just ignoring it like the code in `UserAuthentication.cs`.